pkg/helm/artifacthub_cached.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return nil, err
+	}
+	client := new(http.Client)
+	return &ArtifactHubCachedPackageClient{
+		APIRoot:   apiRoot,
+		URL:       u,
+		Client:    client,
+		UserAgent: fmt.Sprintf("Fairwinds-Nova/%s ", version),
+	}, nil
+}
+
+// List returns all packages from ArtifactHub
+func (ac *ArtifactHubCachedPackageClient) List() ([]ArtifactHubHelmPackage, error) {
+	list := ArtifactHubCachedPackagesList{}
+	if os.Getenv("ARTIFACT_HUB_CACHE_FILE") == "" {
+		resp, err := ac.get("", url.Values{})
+		if err != nil {
+			return nil, err
+		}
+		err = json.NewDecoder(resp.Body).Decode(&list)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		cache, err := ioutil.ReadFile(os.Getenv("ARTIFACT_HUB_CACHE_FILE"))
+		if err != nil {
+			return nil, err
+		}
+		err = json.Unmarshal(cache, &list)
+		if err != nil {
+			return nil, err
+		}
+	}
+	packages := make([]ArtifactHubHelmPackage, len(list))
+	for idx, cachedPackage := range list {
+		packages[idx] = ArtifactHubHelmPackage{
+			Name:        cachedPackage.Name,
+			Description: cachedPackage.Description,
+			Maintainers: cachedPackage.Maintainers,
+			HomeURL:     cachedPackage.HomeURL,
+			Links:       cachedPackage.Links,
+			Official:    cachedPackage.Official,
+			Repository: ArtifactHubRepository{
+				Name:              cachedPackage.Repository.Name,
+				URL:               cachedPackage.Repository.URL,
+				VerifiedPublisher: cachedPackage.Repository.Verified,
+				Official:          cachedPackage.Repository.Official,
+			},
+			Version:           cachedPackage.LatestVersion,
+			AvailableVersions: []AvailableVersion{},
+		}
+		for _, version := range cachedPackage.Versions {
+			if version.Version == cachedPackage.LatestVersion {
+				packages[idx].AppVersion = version.AppVersion
+				packages[idx].Deprecated = version.Deprecated
+			}
+			packages[idx].AvailableVersions = append(packages[idx].AvailableVersions, AvailableVersion{
+				Version: version.Version,
+			})
+		}
+	}
+	return packages, nil
+}
+
+// get is the basic getter for the artifacthub package client
+// The path argument should be formatted like so: "api/v1/packages/search", any unauthenticated paths
+// will work and are documented here: https://artifacthub.io/docs/api/#/
+// urlValues are the search parameters for the query if necessary.
+// offset is to be used for pagination. The first page would be offset 0.
+func (ac *ArtifactHubCachedPackageClient) get(path string, urlValues url.Values) (*http.Response, error) {
+	requestURL := *ac.URL
+	requestURL.Path = path
+	urlString := requestURL.String()
+	r, err := http.NewRequest("GET", urlString, nil)
+	if err != nil {
+		return nil, err
+	}
+	q := r.URL.Query()
+	for k, v := range urlValues {
+		for _, vv := range v {
+			q.Add(k, vv)
+		}
+	}
+	r.URL.RawQuery = q.Encode()
+	r.Header.Add("accept", "application/json")
+	r.Header.Set("User-Agent", ac.UserAgent)
+	var response *http.Response
+	for attempt := 1; attempt <= 5; attempt++ {
```

Comment:
It'd be great if you could use some sort of exponential backoff (like https://github.com/cenkalti/backoff) or introduce a pause instead of retrying immediately a few times when something goes wrong. In some situations (like when encountering rate limit errors), this may make things even worse.

Reply:
Good catch! I removed the retry logic entirely. I think it's an artifact from when we originally queried the repositories themselves, which were very flaky